sdks/java/io/kinesis/src/main/java/org/apache/beam/sdk/io/kinesis/ShardReadersPool.java

```diff
@@ -69,25 +76,37 @@ void start() throws TransientKinesisException {
       shardsMap.put(checkpoint.getShardId(), createShardIterator(kinesis, checkpoint));
     }
     shardIteratorsMap = shardsMap.build();
-    executorService = Executors.newFixedThreadPool(shardIteratorsMap.size());
-    recordsQueue = new LinkedBlockingQueue<>(queueCapacityPerShard * shardIteratorsMap.size());
-    for (final ShardRecordsIterator shardRecordsIterator : shardIteratorsMap.values()) {
-      executorService.submit(new Runnable() {
-
-        @Override
-        public void run() {
-          readLoop(shardRecordsIterator);
-        }
-      });
+    if (!shardIteratorsMap.isEmpty()) {
+      recordsQueue = new ArrayBlockingQueue<>(queueCapacityPerShard * shardIteratorsMap.size());
+      for (final ShardRecordsIterator shardRecordsIterator : shardIteratorsMap.values()) {
+        numberOfRecordsInAQueueByShard.put(shardRecordsIterator.getShardId(), new AtomicInteger());
+        executorService.submit(new Runnable() {
+
+          @Override
+          public void run() {
+            readLoop(shardRecordsIterator);
+          }
+        });
+      }
+    } else {
+      recordsQueue = new LinkedBlockingQueue<>();
```

Comment:
In this case nobody is running readLoop(), correct? Is this object then in a valid state even?

Reply:
Yes, in this case the readLoop() is not executed and actually this pool won’t do anything. I didn’t find a way to tell Beam that the source using this pool will not produce records anymore.